File: src/export-glossika-deck.ts
import { writeFile } from 'fs/promises';
import { z } from 'zod';
import { ankiRequest, NoteInfo } from './anki-connect.js';

// Parse command-line arguments
const args = process.argv.slice(2);
const DECK_NAME = args[0] || 'Glossika-ENJA [2001-3000]';
const OUTPUT_FILE = args[1] || 'glossika_deck_export.csv';

// Type inference from Zod schemas
type NoteInfo = z.infer<typeof NoteInfo>;
type CsvRow = {
  id: string;
  english: string;
  japanese: string;
  ka: string;
  ROM: string;
  explanation: string;
};

/**
 * Converts an array of rows to CSV format.
 */
function rowsToCsv(rows: CsvRow[]): string {
  const fieldNames: (keyof CsvRow)[] = [
    'id',
    'english',
    'japanese',
    'ka',
    'ROM',
    'explanation',
  ];

  // CSV header
  const header = fieldNames.join(',');

  // CSV rows
  const csvRows = rows.map((row) => {
    return fieldNames
      .map((field) => {
        const value = row[field];
        // Escape quotes and wrap in quotes if the value contains comma, quote, or newline
        if (
          value.includes(',') ||
          value.includes('"') ||
          value.includes('\n')
        ) {
          return `"${value.replace(/"/g, '""')}"`;
        }
        return value;
      })
      .join(',');
  });

  return [header, ...csvRows].join('\n');
}

/**
 * Helper function to safely extract and clean field values from a note.
 */
function getFieldValue(fields: NoteInfo['fields'], fieldName: string): string {
  const rawValue = fields[fieldName]?.value ?? '';
  return rawValue.replace(/\r/g, '');
}

async function exportDeckToCsv(): Promise<void> {
  console.log('='.repeat(60));
  console.log(`Exporting deck: ${DECK_NAME}`);
  console.log('='.repeat(60));

  try {
    // Find all notes in the deck
    const query = `deck:"${DECK_NAME}"`;
    const noteIds = await ankiRequest('findNotes', z.array(z.number()), {
      query,
    });
    console.log(`\n✓ Found ${noteIds.length} notes in '${DECK_NAME}'.`);

    if (noteIds.length === 0) {
      console.log('No notes found to export.');
      return;
    }

    // Get detailed info for all notes
    console.log(`\nFetching note details...`);
    const notesInfo = await ankiRequest('notesInfo', z.array(NoteInfo), {
      notes: noteIds,
    });
    console.log(`✓ Retrieved information for ${notesInfo.length} notes.`);

    // Convert notes to CSV rows
    const rows: CsvRow[] = notesInfo.map((note: NoteInfo) => {
      return {
        id: getFieldValue(note.fields, 'Id'),
        english: getFieldValue(note.fields, 'English'),
        japanese: getFieldValue(note.fields, 'Japanese'),
        ka: getFieldValue(note.fields, 'か'),
        ROM: getFieldValue(note.fields, 'ROM'),
        explanation: getFieldValue(note.fields, 'Explanation'),
      };
    });

    // Write to CSV
    console.log(`\nWriting to ${OUTPUT_FILE}...`);
    const csvContent = rowsToCsv(rows);

    await writeFile(OUTPUT_FILE, csvContent, 'utf-8');
    console.log(
      `✓ Successfully exported ${notesInfo.length} notes to ${OUTPUT_FILE}`,
    );
  } catch (error) {
    if (error instanceof Error) {
      console.log(`\n✗ Error: ${error.message}`);
      if (error instanceof z.ZodError) {
        console.log('Validation details:', error.flatten());
      }
    } else {
      console.log('\n✗ An unknown error occurred:', error);
    }
    console.log('\nMake sure:');
    console.log('  1. Anki Desktop is running');
    console.log('  2. AnkiConnect add-on is installed (code: 2055492159)');
    console.log(`  3. Deck '${DECK_NAME}' exists`);
  }
}

void exportDeckToCsv();

File: export_glossika_deck.py
import json
import csv
import requests

ANKI_CONNECT_URL = "http://127.0.0.1:8765"
DECK_NAME = "Glossika-ENJA [2001-3000]"
OUTPUT_FILE = "glossika_deck_export.csv"

def anki_request(action, **params):
    """
    Helper function to send requests to AnkiConnect.
    """
    payload = {"action": action, "params": params, "version": 6}
    try:
        response = requests.post(ANKI_CONNECT_URL, data=json.dumps(payload))
        response.raise_for_status()
        response_json = response.json()
        if response_json.get("error"):
            raise Exception(f"AnkiConnect API error: {response_json['error']}")
        return response_json.get("result")
    except requests.exceptions.RequestException as e:
        raise Exception(f"Could not connect to AnkiConnect. Is Anki running? Error: {e}")

def export_deck_to_csv():
    print("=" * 60)
    print(f"Exporting deck: {DECK_NAME}")
    print("=" * 60)

    try:
        # Find all notes in the deck
        query = f'deck:"{DECK_NAME}"'
        note_ids = anki_request('findNotes', query=query)
        print(f"\n✓ Found {len(note_ids)} notes in '{DECK_NAME}'.")

        if not note_ids:
            print("No notes found to export.")
            return

        # Get detailed info for all notes
        print(f"\nFetching note details...")
        notes_info = anki_request('notesInfo', notes=note_ids)
        print(f"✓ Retrieved information for {len(notes_info)} notes.")

        # Write to CSV
        print(f"\nWriting to {OUTPUT_FILE}...")
        with open(OUTPUT_FILE, 'w', newline='', encoding='utf-8') as csvfile:
            # Define the field names for CSV
            fieldnames = ['id', 'english', 'japanese', 'ka', 'ROM', 'explanation']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames, lineterminator='\n')

            writer.writeheader()

            for note in notes_info:
                fields = note['fields']
                row = {
                    'id': fields.get('Id', {}).get('value', '').replace('\r', ''),
                    'english': fields.get('English', {}).get('value', '').replace('\r', ''),
                    'japanese': fields.get('Japanese', {}).get('value', '').replace('\r', ''),
                    'ka': fields.get('か', {}).get('value', '').replace('\r', ''),
                    'ROM': fields.get('ROM', {}).get('value', '').replace('\r', ''),
                    'explanation': fields.get('Explanation', {}).get('value', '').replace('\r', '')
                }
                writer.writerow(row)

        print(f"✓ Successfully exported {len(notes_info)} notes to {OUTPUT_FILE}")

    except Exception as e:
        print(f"\n✗ Error: {e}")
        print("\nMake sure:")
        print("  1. Anki Desktop is running")
        print("  2. AnkiConnect add-on is installed (code: 2055492159)")
        print(f"  3. Deck '{DECK_NAME}' exists")

if __name__ == "__main__":
    export_deck_to_csv()
